lib/honeybadger/check_in.rb

```diff
@@ -0,0 +1,87 @@
+module Honeybadger
+  class InvalidCheckinConfig < StandardError; end
+  class CheckInSyncError < StandardError; end
+  
+  class CheckIn
+    attr_reader :project_id
+    attr_accessor :name, :slug, :schedule_type, :report_period, :grace_period, :cron_schedule, :cron_timezone, :deleted, :id
+    def self.from_config(config_entry)
+      c = config_entry.transform_keys {|k| k.to_s }
+      
+      check_in = self.new(c["project_id"], c["id"])
+      check_in.name = c["name"]
+      check_in.slug = c["slug"]
+      check_in.schedule_type = c["schedule_type"]
+      check_in.report_period = c["report_period"]
+      check_in.grace_period = c["grace_period"]
+      check_in.cron_schedule = c["cron_schedule"]
+      check_in.cron_timezone = c["cron_timezone"]
+      check_in
+    end
+    
+    def self.from_remote(project_id, data)
+      check_in = self.new(project_id, data["id"])
+      check_in.name = data["name"]
+      check_in.slug = data["slug"]
+      check_in.schedule_type = data["schedule_type"]
+      check_in.report_period = data["report_period"]
+      check_in.grace_period = data["grace_period"]
+      check_in.cron_schedule = data["cron_schedule"]
+      check_in.cron_timezone = data["cron_timezone"]
+      check_in
+    end
+
+    def initialize(project_id, id = nil)
+      @project_id = project_id
+      @id = id
+      @deleted = false
+    end
+
+    def ==(other)
+      self.name == other.name &&
+      self.slug == other.slug &&
+      self.schedule_type == other.schedule_type &&
+      self.report_period == other.report_period &&
+      self.grace_period == other.grace_period &&
+      self.cron_schedule == other.cron_schedule &&
+      self.cron_timezone == other.cron_timezone
+    end
+
+    def to_json
```

Comment:
Is there any chance you are going to be sending `nil` values for the optional properties? If that's the case, they should be changed to empty strings.
(@stympy correct me if I'm wrong)

Reply:
They should be nils, not empty strings.

Reply:
I will test the empty strings! If I understood well from our [conversation last time](https://honeybadger.slack.com/archives/C024F9JHC/p1696003452553299?thread_ts=1695989539.033289&cid=C024F9JHC), sending `nil` resulted in errors.

That's what led me to ignore optional values if they are not set (i.e. don't send them at all), but this creates an issues:
- If an optional field, such as `slug` is set, and now we want to unset it, it won't work because we won't be sending the `nil` value.

I think our best course of action here is to create constants of the defaults on the server and use them instead of `nil` (or empty strings).

Reply:
Thanks for the reminder! Yes, sending nulls for fields that have a no-null constraint won't work very well. :) I'm agreed on not sending optional values that aren't set.

> If an optional field, such as slug is set, and now we want to unset it, it won't work because we won't be sending the nil value.

We have special treatment for slug on the server-side... we nullify empty strings for that field, so you can send an empty string to unset the slug.

> I think our best course of action here is to create constants of the defaults on the server and use them instead of nil (or empty strings).

This sounds reasonable to me.

Reply:
The default values:
- `slug` -> `null` or `''`
- `grace_period` -> `00:00:00` or `''`

Reply:
Okay, so what I haven't really gathered from your conversation is if I should implement this now in this PR or if changes to handle this in the API are under way?

Reply:
Nevermind: I implemented roughly what's implemented in the PHP implementation but I wonder if I should be more selective with the payload (as in: only add the elements to the payload relevant for the `schedule_type`)

Reply:
> I wonder if I should be more selective with the payload (as in: only add the elements to the payload relevant for the schedule_type)

That would be preferred

Reply:
It might be best to send all the keys... I'm thinking of the scenario where someone changes from the simple schedule to the cron schedule or vice-versa. In those cases it could make sense to send them all in order to reset the now-unused values on the server side.

Reply:
> It might be best to send all the keys... I'm thinking of the scenario where someone changes from the simple schedule to the cron schedule or vice-versa.

Currently I'm not sending all the keys, only because I think it's not necessary (that's what I understood from looking at the backend).
But yeah, we could do some testing. @halfbyte can you please try changing a check-in from simple to cron and vice-verca?

Reply:
> Currently I'm not sending all the keys, only because I think it's not necessary (that's what I understood from looking at the backend).

True the backend _should_ be taking care of it. :)

Reply:
From what I can see, the backend handles missing keys just fine and also seems to reset fields for the "inactive" schedule type properly. If I change a simple check in to type "cron" with a proper `cron_schedule` and then try to change it back to "simple" without also setting `report_period` the server throws an error.

I'll adapt my code to closer follow the PHP implementation, I think?

Reply:
> From what I can see, the backend handles missing keys just fine and also seems to reset fields for the "inactive" schedule type properly. If I change a simple check in to type "cron" with a proper cron_schedule and then try to change it back to "simple" without also setting report_period the server throws an error.

This is the correct behavior though, right?
I would expect that you would receive an error before even reaching the server, when the check-ins are loaded from the config file and [validated](https://github.com/honeybadger-io/honeybadger-ruby/pull/508/files#diff-2306255ad6db4971c5877c2f3b98218488fca0fb02f6b95895df8b65a604c5e4R59).

> I'll adapt my code to closer follow the PHP implementation, I think?

What would you change here? I thought that this part follows the PHP implementation as it is now.